vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java

```diff
@@ -0,0 +1,188 @@
+public class JSONParser {
+
+    final JsonFactory factory;
+    protected final ArrayBackedValueStorage atomic;
+    protected final List<ArrayBuilder> abStack;
+    protected final List<ObjectBuilder> obStack;
+    protected final List<ArrayBackedValueStorage> abvsStack;
+    protected final List<UTF8StringPointable> spStack;
+    protected final List<String> itemStack;
+
+    public JSONParser() throws JsonParseException, IOException {
+        factory = new JsonFactory();
+        atomic = new ArrayBackedValueStorage();
+        abStack = new ArrayList<ArrayBuilder>();
+        obStack = new ArrayList<ObjectBuilder>();
+        abvsStack = new ArrayList<ArrayBackedValueStorage>();
+        spStack = new ArrayList<UTF8StringPointable>();
+        itemStack = new ArrayList<String>();
+    }
+
+    public void parseDocument(File file, ArrayBackedValueStorage result)
+            throws NumberFormatException, JsonParseException, IOException {
+        DataOutput outResult = result.getDataOutput();
+        DataOutput out = atomic.getDataOutput();
+        StringValueBuilder svb = new StringValueBuilder();
+        String lastItem = null;
+        String startItem = null;
+        boolean nested = false;
+        JsonParser parser = factory.createParser(file);
+        while (!parser.isClosed()) {
+            JsonToken token = parser.nextToken();
+            if (token == JsonToken.START_ARRAY) {
+                if (!nested) {
+                    abvsStack.add(new ArrayBackedValueStorage());
+                    abStack.add(new ArrayBuilder());
+                }
+                itemStack.add("array");
+                lastItem = "array";
+                abvsStack.get(abvsStack.size() - 1).reset();
+                abStack.get(abStack.size() - 1).reset(abvsStack.get(abvsStack.size() - 1));
+            }
+            if (token == JsonToken.START_OBJECT) {
+                if (!nested) {
+                    abvsStack.add(new ArrayBackedValueStorage());
+                    obStack.add(new ObjectBuilder());
+                }
+                itemStack.add("object");
+                lastItem = "object";
+                abvsStack.get(abvsStack.size() - 1).reset();
+                obStack.get(obStack.size() - 1).reset(abvsStack.get(abvsStack.size() - 1));
+            }
+            if (token == JsonToken.FIELD_NAME) {
+                UTF8StringPointable sp = new UTF8StringPointable();
+                ArrayBackedValueStorage key = new ArrayBackedValueStorage();
+                key.reset();
+                spStack.add(sp);
+                DataOutput outk = key.getDataOutput();
+                svb.write(parser.getText(), outk);
+                spStack.get(spStack.size() - 1).set(key);
+            }
+            if (token == JsonToken.VALUE_NUMBER_INT) {
+                atomic.reset();
+                out.write(ValueTag.XS_INTEGER_TAG);
+                out.writeLong(parser.getLongValue());
+                if (lastItem == "array") {
```

Comment:
We could create a method for this if-then-else that could be called for all atomic values.